node/bindnode/node.go

```diff
@@ -625,22 +709,31 @@ func (w *_assembler) AssignBytes(p []byte) error {
 }
 
 func (w *_assembler) AssignLink(link datamodel.Link) error {
-	newVal := reflect.ValueOf(link)
-	if !newVal.Type().AssignableTo(w.val.Type()) {
-		if newVal.Type() == goTypeCidLink && goTypeCid.AssignableTo(w.val.Type()) {
-			// Unbox a cidlink.Link to assign to a go-cid.Cid value.
-			newVal = newVal.FieldByName("Cid")
-		} else {
-			// The target value cannot be assigned a datamodel.Link or go-cid.Cid.
-			return datamodel.ErrWrongKind{
-				TypeName:        w.schemaType.Name(),
-				MethodName:      "AssignLink",
-				AppropriateKind: datamodel.KindSet_JustLink,
-				ActualKind:      w.kind(),
+	if _, ok := w.schemaType.(*schema.TypeAny); ok {
+		basicBuilder := basicnode.Prototype.Any.NewBuilder()
+		if err := basicBuilder.AssignLink(link); err != nil {
+			return err
+		}
+		basicNode := basicBuilder.Build()
+		w.nonPtrVal().Set(reflect.ValueOf(basicNode))
+	} else {
+		newVal := reflect.ValueOf(link)
+		if !newVal.Type().AssignableTo(w.val.Type()) {
+			if newVal.Type() == goTypeCidLink && goTypeCid.AssignableTo(w.val.Type()) {
+				// Unbox a cidlink.Link to assign to a go-cid.Cid value.
+				newVal = newVal.FieldByName("Cid")
+			} else {
+				// The target value cannot be assigned a datamodel.Link or go-cid.Cid.
+				return datamodel.ErrWrongKind{
+					TypeName:        w.schemaType.Name(),
+					MethodName:      "AssignLink",
+					AppropriateKind: datamodel.KindSet_JustLink,
+					ActualKind:      actualKind(w.schemaType),
+				}
```

Comment:
This error message doesn't seem right.  I think a custom message might be needed here.  In this branch, we're dealing: EITHER the field on the value we're reflecting on isn't something that bindnode understands as a link type, OR, bindnode might understand it as a link type, but the parameter we were given isn't go-cid.Cid.  Is that understanding correct?  The former case sorta fits ErrWrongKind; the latter doesn't at all.

(It may have been equally dubious before this diff too; I'm just catching it now.)

Reply:
Kinda suspect this whole area might be easier to reason about if it was rewritten to use earlier returns and have less leftward drift.  E.g.,

```
// If both the field and the parameter are `datamodel.Link`, easy: assign directly.
if newVal.Type().AssignableTo(w.val.Type()) {
    break
}
// ... more checks below, each breaking out as soon as possible ...
```

Reply:
Added a TODO for now, as this code was already here - this is just an indentation change. I'd rather skip this for now as I want to get graphsync working before blocking Hannah and Rod for another two weeks.